File: sdk/python/tempmail_sdk/providers/one_sec_mail.py
"""
1SecMail 渠道 -- https://tmaily.com
"""

from typing import Any, Dict, List

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "1sec-mail"
BASE_URL = "https://tmaily.com/"


def _extract_cookie(resp: Any) -> str:
    """从响应中提取 TMaily_sid Cookie"""
    for cookie in resp.cookies:
        if cookie.name == "TMaily_sid":
            return f"TMaily_sid={cookie.value}"
    return ""


def generate_email() -> EmailInfo:
    """生成临时邮箱，从 Set-Cookie 提取会话 Cookie"""
    resp = tm_http.get(
        f"{BASE_URL}generate",
        headers={"Accept": "application/json", "User-Agent": "Mozilla/5.0"},
        timeout=15,
    )
    resp.raise_for_status()
    cookie = _extract_cookie(resp)
    if not cookie:
        raise ValueError("1sec-mail: 未获取到会话 Cookie")
    data = resp.json()
    if not isinstance(data, dict) or not isinstance(data.get("address"), str):
        raise ValueError("1sec-mail: 无效的邮箱响应")
    email = data["address"].strip()
    if "@" not in email:
        raise ValueError("1sec-mail: 无效的邮箱响应")
    return EmailInfo(channel=CHANNEL, email=email, _token=cookie)


def get_emails(token: str, email: str) -> List[Email]:
    """轮询邮件列表，带上会话 Cookie"""
    address = (email or "").strip()
    if not address:
        raise ValueError("1sec-mail: 邮箱地址为空")
    resp = tm_http.get(
        f"{BASE_URL}emails?address={address}",
        headers={
            "Accept": "application/json",
            "Cookie": token,
            "User-Agent": "Mozilla/5.0",
        },
        timeout=15,
    )
    resp.raise_for_status()
    rows = resp.json()
    if not isinstance(rows, list):
        raise ValueError("1sec-mail: 无效的邮件列表响应")
    return [normalize_email(item, address) for item in rows if isinstance(item, dict)]

File: sdk/ruby/lib/tempmail_sdk/providers/one_sec_mail.rb
# frozen_string_literal: true

module TempmailSdk
  module Providers
    # 1SecMail 渠道 — https://tmaily.com
    module OneSecMail
      CHANNEL = "1sec-mail"
      BASE_URL = "https://tmaily.com/"
      HEADERS = { "Accept" => "application/json", "User-Agent" => "Mozilla/5.0" }.freeze

      module_function

      # 生成临时邮箱，从 Set-Cookie 提取会话 Cookie
      # @return [EmailInfo]
      def generate_email
        resp = Http.get("#{BASE_URL}generate", headers: HEADERS, timeout: 15)
        resp.raise_for_status
        cookie_val = resp.cookie_value("TMaily_sid")
        raise "1sec-mail: 未获取到会话 Cookie" if cookie_val.nil? || cookie_val.empty?

        cookie = "TMaily_sid=#{cookie_val}"
        data = resp.json
        email = data.is_a?(Hash) ? data["address"].to_s.strip : ""
        raise "1sec-mail: 无效的邮箱响应" unless email.include?("@")

        EmailInfo.new(channel: CHANNEL, email: email, token: cookie)
      end

      # @param email [String]
      # @param token [String] 会话 Cookie
      # @return [Array<Email>]
      def get_emails(email, token)
        address = email.to_s.strip
        raise "1sec-mail: 邮箱地址为空" if address.empty?

        resp = Http.get("#{BASE_URL}emails?address=#{address}",
                        headers: HEADERS.merge("Cookie" => token.to_s), timeout: 15)
        resp.raise_for_status
        rows = resp.json
        raise "1sec-mail: 无效的邮件列表响应" unless rows.is_a?(Array)

        rows.select { |r| r.is_a?(Hash) }.map { |item| Normalize.normalize_email(item, address) }
      end
    end
  end
end
